pinot-connectors/pinot-spark-connector/src/main/scala/org/apache/pinot/connector/spark/connector/PinotClusterClient.scala

```diff
@@ -0,0 +1,208 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.connector.spark.connector
+
+import java.net.{URI, URLEncoder}
+import java.util.regex.Pattern
+
+import org.apache.pinot.connector.spark.decodeTo
+import org.apache.pinot.connector.spark.exceptions.{HttpStatusCodeException, PinotException}
+import org.apache.pinot.connector.spark.utils.{HttpUtils, Logging}
+
+import scala.util.{Failure, Success, Try}
+import io.circe.generic.auto._
+import org.apache.pinot.connector.spark.connector.Constants.PinotTableTypes
+import org.apache.pinot.connector.spark.connector.query.GeneratedPQLs
+import org.apache.pinot.spi.data.Schema
+
+/**
+ * Client that read/write/prepare required data from/to Pinot.
+ */
+private[pinot] object PinotClusterClient extends Logging {
+
+  def getTableSchema(controllerUrl: String, tableName: String): Schema = {
+    val rawTableName = PinotUtils.getRawTableName(tableName)
+    Try {
+      val uri = new URI(s"http://$controllerUrl/tables/$rawTableName/schema")
+      val response = HttpUtils.sendGetRequest(uri)
+      Schema.fromString(response)
+    } match {
+      case Success(response) =>
+        logDebug(s"Pinot schema received successfully for table '$rawTableName'")
+        response
+      case Failure(exception) =>
+        throw PinotException(
+          s"An error occurred while getting Pinot schema for table '$rawTableName'",
+          exception
+        )
+    }
+  }
+
+  /**
+   * Get available broker urls(host:port) for given table.
+   * This method is used when if broker instances not defined in the datasource options.
+   */
+  def getBrokerInstances(controllerUrl: String, tableName: String): List[String] = {
+    val brokerPattern = Pattern.compile("Broker_(.*)_(\\d+)")
+    val rawTableName = PinotUtils.getRawTableName(tableName)
+    Try {
+      val uri = new URI(s"http://$controllerUrl/tables/$rawTableName/instances")
```

Comment:
I think we have a new Controller API that provides this as one call cc @fx19880617

Reply:
i'll check it. Connector was written with pinot 0.4.0. I can update it